lib/shopify_app/controller_concerns/token_exchange.rb

```diff
@@ -61,23 +68,28 @@ def id_token_header
     end
 
     def respond_to_invalid_shopify_id_token
-      # TODO: Implement this method to handle invalid session tokens
+      return redirect_to_bounce_page if request.headers["HTTP_AUTHORIZATION"].blank?
 
-      # if request.xhr?
-      # response.set_header("X-Shopify-Retry-Invalid-Session-Request", 1)
-      # unauthorized_response = { message: :unauthorized }
-      # render(json: { errors: [unauthorized_response] }, status: :unauthorized)
-      # else
-      # patch_session_token_url = "#{ShopifyAPI::Context.host}/patch_session_token"
-      # patch_session_token_params = request.query_parameters.except(:id_token)
+      ShopifyApp::Logger.debug("Responding to invalid Shopify ID token with unauthorized response")
+      response.set_header("X-Shopify-Retry-Invalid-Session-Request", 1)
+      unauthorized_response = { message: :unauthorized }
+      render(json: { errors: [unauthorized_response] }, status: :unauthorized)
+    end
+
+    def redirect_to_bounce_page
+      ShopifyApp::Logger.debug("Redirecting to bounce page for patching Shopify ID token")
+      patch_shopify_id_token_url = "#{ShopifyApp.configuration.root_url}/patch_shopify_id_token"
+      patch_shopify_id_token_params = request.query_parameters.except(:id_token)
 
-      # bounce_url = "#{ShopifyAPI::Context.host}#{request.path}?#{patch_session_token_params.to_query}"
+      bounce_url = "#{request.path}?#{patch_shopify_id_token_params.to_query}"
 
-      # # App Bridge will trigger a fetch to the URL in shopify-reload, with a new session token in headers
-      # patch_session_token_params["shopify-reload"] = bounce_url
+      # App Bridge will trigger a fetch to the URL in shopify-reload, with a new session token in headers
+      patch_shopify_id_token_params["shopify-reload"] = bounce_url
 
-      # redirect_to("#{patch_session_token_url}?#{patch_session_token_params.to_query}", allow_other_host: true)
-      # end
+      redirect_to(
+        "#{patch_shopify_id_token_url}?#{patch_shopify_id_token_params.to_query}",
```

Comment:
I wish there was a way to use the Rails URL helper method `patch_shopify_id_token_path` here which already does most of the URL assembly work for us, but the way our tests are setup using `with_routing` it's not including engine's routes, so I think it would be a bit too much work for it to be worth it.